Document `pub` items in multiple `clippy_utils` modules

SignalWalker opened the pull request:
Partially addressing rust-lang/rust-clippy#15569, document items in `clippy_utils::{consts, higher, macros, paths, res, source, str_utils, sugg, usage, visitors}`. The specific items documented are listed in each commit description, along with notes.

This is my first time contributing to any official Rust internals/tools, so please let me know if I misunderstood anything. (I'm not really sure that "refers" is the right verb for the relationship between a `Place` and a HIR node, for example.)

As I understand it, this is considered purely an internal change, so:
changelog: none

SignalWalker committed:
Document `pub` items in `clippy_utils::usage`

- `is_potentially_mutated()`
- `is_potentially_local_place()`
- `ParamBindingIdCollector`
- `BindingUsageFinder`
- `contains_return_break_continue_macro()`
- `local_used_in()`
- `local_used_after_expr()`

SignalWalker committed:
Document `pub` items in `clippy_utils::higher`

- `IfLetOrMatch::scrutinee()`
- `While.label`
- `WhileLet.label`

SignalWalker committed:
Document `pub` items in `clippy_utils::macros`

- module
- `MacroCall::is_local()`
- `PanicExpn`
- `PanicExpn::parse()`
- `HirNode::hir_id()`
- `HirNode::span()`

SignalWalker committed:
Document `clippy_utils::paths::PathNS::{Type, Value, Macro}`

I'm not really sure how to document the `PathLookup` statics,
so I'm skipping those for now.

SignalWalker committed:
Document `pub` items in `clippy_utils::res`

- module
- `HasHirId::hir_id()`
- `MaybeTypeckRes`
- `MaybeDef::opt_def_id()`

SignalWalker committed:
Document `pub` items in `clippy_utils::source`

- `HasSession`
- `HasSession::sess()`
- `SpanRange::into_range()`
- `IntoSpan::into_span()`
- `IntoSpan::with_ctxt()`
- `SpanRangeExt`
  - May want to be more specific with this one.
- `SourceFileRange` and its fields
- `is_present_in_source()`
  - This one feels a bit clunky to me; might want to rephrase.
- `snippet_block_with_context()`

SignalWalker committed:
Document `pub` items in `clippy_utils::str_utils`

- module
- fields of `StrIndex`
- `StrIndex::new()`
- fields of `StrCount`
- `StrCount::new()`

`StrIndex` and `StrCount` don't seem to guarantee that their fields are
internally consistent; should this be documented?

SignalWalker committed:
Document `pub` items in `clippy_utils::sugg`

- `Sugg::into_string()`

I'm not sure that this is correct, but it does seem to be how its used
elsewhere.

SignalWalker committed:
Document `pub` items in `clippy_utils::visitors`

- module
- `Descend::{Yes, No}`
- `find_all_ret_expressions()`
- `for_each_unconsumed_temporary()`
  - this one was mostly just reformatting the non-doc comment that was
    already there
- `any_temporaries_need_ordered_drop()`
- `contains_break_or_continue()`

SignalWalker committed:
Document `pub` items in `clippy_utils::consts`

- `Constant::Adt`
- `Constant::partial_cmp()`
- `Constant::peel_refs()`
- `Constant::new_numeric_min()`
- `Constant::new_numeric_max()`
- `Constant::is_numeric_min()`
- `Constant::is_numeric_max()`
- `Constant::is_pos_infinity()`
- `Constant::is_neg_infinity()`
- `ConstantSource::is_local()`
- `FullInt` and its variants
- `ConstEvalCtxt::eval_pat_expr()`
- `mir_to_const()`

llogiq commented on clippy_utils/src/consts.rs:
```diff
@@ -28,6 +28,7 @@ use std::iter;
 /// A `LitKind`-like enum to fold constant `Expr`s into.
 #[derive(Debug, Clone)]
 pub enum Constant {
+    /// A constant representing an algebraic data type
```
Interesting, This could still change in the future, so the comment is fine here.

llogiq commented on clippy_utils/src/consts.rs:
```diff
@@ -272,6 +282,7 @@ impl Constant {
         }
     }
 
+    /// Recursively consume [`Constant::Ref`] and return the innermost value.
```
How about "removing all references to return the contained expression, e.g. &&T → T"?

llogiq commented on clippy_utils/src/consts.rs:
```diff
@@ -447,14 +468,20 @@ pub enum ConstantSource {
 }
 
+/// An integer type (signed or unsigned) with enough bits to represent any integer that can be
+/// represented in Rust.
```
Indeed.

llogiq commented on clippy_utils/src/macros.rs:
```diff
@@ -228,6 +231,7 @@ pub fn is_assert_macro(cx: &LateContext<'_>, def_id: DefId) -> bool {
     matches!(name, sym::assert_macro | sym::debug_assert_macro)
 }
 
+/// A `panic!()` expression, which may contain arguments
```
`Expn` stands for *expansion* here, not *expression*. So it refers to the expansion of the `panic!`  macro.

llogiq commented on clippy_utils/src/source.rs:
```diff
@@ -631,6 +645,8 @@ pub fn snippet_block_with_applicability(
 }
 
+/// Same as [`snippet_block_with_applicability()`], but first walks the span up to the given context
+/// using [`snippet_with_context()`].
```
Yes, please. I for one wouldn't understand what this does without following the link.

llogiq commented on clippy_utils/src/str_utils.rs:
```diff
@@ -169,11 +174,14 @@ pub fn camel_case_split(s: &str) -> Vec<&str> {
 /// character and byte count are provided for correct indexing.
```
If we already have a note at the beginning of the file, that shouldn't be necessary.

llogiq commented on clippy_utils/src/usage.rs:
```diff
@@ -181,6 +191,7 @@ pub fn contains_todo_unimplement_macro(cx: &LateContext<'_>, expr: &'_ Expr<'_>)
     .is_some()
 }
 
+/// Checks whether the given expression contains `return`, `break`, `continue`, or a macro call.
```
Probably because of `panic!` et al.

llogiq reviewed (commented):
Looks mostly good. I left a few suggestions.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/15855/d9fb15c4b1ebe9e7dc419e07f53af681d7860cbe..07d0db91d8110b183f2d26e1d9b2c45337b3b8bb)*

SignalWalker committed:
Apply suggestion

Co-authored-by: llogiq <bogusandre@gmail.com>

SignalWalker committed:
Apply doc suggestions

- Clarify:
  - `consts::peel_refs()`
  - `consts::mir_to_const()`
  - `source::snippet_block_with_context()`
  - `str_utils::StrIndex`
  - `str_utils::StrCount`
  - `sugg::Sugg::into_string()`
  - `usage`
  - `usage::contains_return_break_continue_macro()`
- Fix:
  - `macros::PanicExpn`

SignalWalker commented:
Applied your suggestions; thanks!

@rustbot ready

llogiq commented on clippy_utils/src/usage.rs:
```diff
@@ -29,10 +31,13 @@ pub fn mutated_variables<'tcx>(expr: &'tcx Expr<'_>, cx: &LateContext<'tcx>) ->
     mutated_variables(expr, cx).is_none_or(|mutated| mutated.contains(&variable))
 }
 
+/// Checks whether it is possible that the given [Place] refers to the given local.
```
One could also say "is".

llogiq commented:
Sorry for letting this wait for so long. I think it's a solid improvement already, so if you're up to, let's rebase and merge, and we can improve on that in a followup PR later.
